addons/pos_mercado_pago/controllers/main.py

```diff
@@ -0,0 +1,75 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+import hashlib
+import hmac
+import logging
+import re
+
+from odoo import http
+from odoo.http import request
+
+_logger = logging.getLogger(__name__)
+
+
+class PosMercadoPagoWebhook(http.Controller):
+    @http.route('/pos_mercado_pago/notification', methods=['POST'], type="http", auth="none", csrf=False)
+    def notification(self):
+        """ Process the notification sent by Mercado Pago
+
+        Notification format is always json
+        """
+        _logger.debug('POST message received on the end point')
+
+        # Check for mandatory keys in header
+        x_request_id = request.httprequest.headers.get('X-Request-Id')
+        x_signature = request.httprequest.headers.get('X-Signature')
+        if not x_request_id or not x_signature:
+            _logger.debug('POST message received with no X-Request-Id and/or no X-Signature in header')
+            return http.Response(status=400)
+
+        # Check for payload
+        data = request.httprequest.get_json(silent=True)
+        if not data:
+            _logger.debug('POST message received with no data')
+            return http.Response(status=400)
+
+        # If and only if this webhook is related with a payment intend (see payment_mercado_pago.js)
+        # then the field data['additional_info']['external_reference'] contains a string
+        # formated like "XXX_YYY" where "XXX" is the session_id and "YYY" is the payment_method_id
+        external_reference = data.get('additional_info', {}).get('external_reference')
+
+        if not external_reference or not re.fullmatch(r'\d+_\d+', external_reference):
+            _logger.debug('POST message received with no or malformed "external_reference" key')
+            return http.Response(status=400)
+
+        session_id, payment_method_id = external_reference.split('_')
+
+        pos_session_sudo = request.env['pos.session'].sudo().browse(int(session_id))
+        if not pos_session_sudo or pos_session_sudo.state != 'opened':
+            _logger.error("Invalid session id: %s", session_id)
+            # This error is not related with Mercado Pago, simply acknowledge Mercado Pago message
+            return http.Response('OK', status=200)
+
+        payment_method_sudo = pos_session_sudo.config_id.payment_method_ids.filtered(lambda p: p.id == int(payment_method_id))
+        if not payment_method_sudo or payment_method_sudo.use_payment_terminal != 'mercado_pago':
+            _logger.error("Invalid payment method id: %s", payment_method_id)
+            # This error is not related with Mercado Pago, simply acknowledge Mercado Pago message
+            return http.Response('OK', status=200)
+
+        # We have to check if this comes from Mercado Pago with the secret key
+        secret_key = payment_method_sudo.mp_webhook_secret_key
+        sig = x_signature.split(',')
+        ts = sig[0].split('=')[1]
+        v1 = sig[1].split('=')[1]
```

Comment:
Indeed, it's the case now but since it's not guaranteed by Mercado Pago, I'll use a regex (already imported now) verification:
```
        ts_m = re.search(r"ts=(\d+)", x_signature)
        v1_m = re.search(r"v1=([a-f0-9]+)", x_signature)
        ts = ts_m.group(1) if ts_m else None
        v1 = v1_m.group(1) if v1_m else None
        if not ts or not v1:
            _logger.debug('Webhook bad X-Signature, ts: %s, v1: %s', ts, v1)
            return http.Response(status=400)
```
Thank you